lisa/core/node.py

```diff
@@ -35,17 +41,55 @@ def createNode(
 
     def setConnectionInfo(
         self,
-        address: str = "",
-        port: int = 22,
-        publicAddress: str = "",
-        publicPort: int = 22,
-        username: str = "root",
-        password: str = "",
-        privateKeyFile: str = "",
-    ):
-        self.connection = SshConnection(
-            address, port, publicAddress, publicPort, username, password, privateKeyFile
-        )
+        address: Optional[object],
+        port: Optional[object],
+        publicAddress: Optional[object],
+        publicPort: Optional[object],
+        username: Optional[object],
+        password: Optional[object],
+        privateKeyFile: Optional[object],
+    ) -> None:
+        if self.connection is not None:
+            raise Exception(
+                "node is set connection information already, cannot set again"
+            )
+        parameters: Dict[str, object] = dict()
+        if address is not None:
+            parameters["address"] = address
+        if port is not None:
+            parameters["port"] = port
+        if publicAddress is not None:
+            parameters["publicAddress"] = publicAddress
+        if publicPort is not None:
+            parameters["publicPort"] = publicPort
+        if username is not None:
+            parameters["username"] = username
+        if password is not None:
+            parameters["password"] = password
+        if privateKeyFile is not None:
+            parameters["privateKeyFile"] = privateKeyFile
+        self.connection = SshConnection(**parameters)
+
+    def execute(self, cmd: str) -> ExecutableResult:
+        result: ExecutableResult
+        cmd_id = random.randint(0, 10000)
+        start_timer = timer()
+        log.debug("remote(%s) cmd[%s] %s", self.isRemote, cmd_id, cmd)
+        if self.isRemote:
+            # remote
+            if self.connection is None:
+                raise Exception("remote node has no connection info")
+            result = self.connection.execute(cmd)
+        else:
+            # local
+            process = Process()
+            with process:
+                process.start(cmd)
+                result = process.waitResult()
+        end_timer = timer()
+        log.info("cmd[%s] executed with %f", cmd_id, end_timer - start_timer)
+        return result
 
-    def connect(self):
-        pass
+    def cleanup(self) -> None:
+        if self.connection is not None:
+            self.connection.cleanup()
```

Comment:
Show me an example? And I saw some documents, it says destructor happens very late. But context manager of __enter__, __exit__ doesn't apply to long life objects. So I have to write it myself.

Reply:
I had to dig into this...I'm coming from the world of C++14 so I assumed destructors were called when objects left their scope, apparently Python is weird and does not do this 😅

What we want to do is have the initialize/destruction of the objects tied directly to their lifetimes. Could we architect this such that `Nodes` are used as context managers? That may actually make sense. But we definitely want to ensure we design around the lifetimes of these objects as best as possible.

_If_ we can't do a context manager (and I think we probably can, though it may take a bit of re-design), from what I can tell the Pythonic term for these non-destructor destructor functions is `close()` instead of `cleanup()`.

Reply:
Renamed for now, we should revisit though.

Reply:
We need to revisit life cycle of each instance, and use context manager as much as possible.

Reply:
Added to backlog
`Revisit life cycle of each action, and other objects hold resource. Use context manager to make sure it's recycled.`